---
title: "Rcwl Pipelines"
author: "Qiang Hu"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Rcwl Pipelines}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Installation
1. Install the package from Bioconductor.

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("RcwlPipelines")
```
The development version is also available to download from Github. 
```{r getDevel, eval=FALSE}
BiocManager::install("hubentu/RcwlPipelines")
```

2. Load the package into the R session.
```{r Load, message=FALSE}
library(RcwlPipelines)
```

# Pipelines summary
There are mainly 4 pipelines are collected in this package. Here are breif introductions to these pipelines. More pipelines and tools are expected to be included in the future.
## DNASeq alignment pipeline
The pipeline can be used to preprocess DNA sequences in fastq format. It can take paired fastqs, read groups from multiple batches as input.
```{r}
inputs(alignMerge)
```

The pipeline includes two steps and several jobs will be run in each step.

1. ``r names(runs(alignMerge))[[1]]``: bwa alignment by read groups.
```{r}
runs(runs(alignMerge)[[1]])
```
	
   * `bwa`: To align fastqs and read groups to reference genome with `bwa`.
   * `sam2bam`: To convert the alignments in "sam" format to "bam" format with `samtools`.
   * `sortBam`: To sort the "bam" file by coordinates with `samtools`.
   * `idxBam`: To index "bam" file with `samtools`.
  
2. ``r names(runs(alignMerge))[[2]]``: Merge by samples and markduplicates.
```{r}
runs(runs(alignMerge)[[2]])
```

   * `mergeBam`: To merge bam files from multiple batches with `picard`.
   * `markdup`: To mark duplicates with `picard`.
   * `samtools_index`: To index bam file with `samtools`.
   * `samtools_flagstat`: To summarize flags in bam with `samtools`.

The final bam files after duplicates marked, bam index, duplicates matrix, and flag statistics summary will be in the output folder.
```{r}
outputs(alignMerge)
```

Here you can find an example to run the pipeline.

<https://hubentu.github.io/others/Rcwl_DNASeq_Align.html>

## RNASeq pipeline
The pipeline was built with reads quality summary, `STAR` alignment, quantification by `featureCounts` and `RSeQC` quality control. Here are the inputs.
```{r}
inputs(rnaseq_Sf)
```

The pipeline includes 6 steps.

* `fastqc`: To run quality summary for raw fastqs with `fastqc`.
* `STAR`: To align fastqs with `STAR`.
* `samtools_index`: To index aligned bam file.
* `samtools_flagstat`: To summary alignment flags.
* `featureCounts`: To quantify gene abundances.
* `RSeQC`: Several steps included.\
	- `gtfToGenePred`: To convert GTF annotation to "genePred" format.
	- `genePredToBed`: To convert "genePred" annotation to "bed" format.
	- `r_distribution`: To run reads distribution over genome features.
	- `gCoverage`: To summarize read coverage over gene body.

The outputs and logs from alignment, quantification and QC steps are collected together to the output folder. A final QC report could be generated by `multiqc`, which is also available in the data package.

An example to run the pipeline.

<https://hubentu.github.io/others/Rcwl_RNASeq.html>

## MC3 somatic variant calling pipeline
The Multi-Center Mutation Calling in Multiple Cancers project (MC3) pipeline was developed by the TCGA group, which was implemented with CWL and available at <https://github.com/OpenGenomics/mc3>. Two major steps, variant calling and merging VCFs to MAF, was imported to the `mc3` pipeline in this package.

The pipeline requires a paired of tumor/normal BAM files and genomic annotation files as inputs.
```{r}
inputs(mc3)
```

Two steps are included.\
* `call_variants`: To call variants by 7 pipelines.
* `covert`: To merge VCFs and convert to MAF

The merged VCF and coverted MAF files will be collected to the output folder.
```{r}
outputs(mc3)
```

Here is an examples to run the pipeline.\
<https://hubentu.github.io/others/Rcwl_MC3.html>

## GATK4 germline variant calling pipeline
The GATK 4 best practice pipeline for gemline variant calling was implemented with Workflow Description Language (WDL). We wrapped the WDL pipeline into 3 steps with `Rcwl`. Here is the details of the pipeline: <https://software.broadinstitute.org/gatk/best-practices/workflow?id=11145>

1. `GAlign` GATK alignment.
The fastqs, sample information and customized json files for WDL are required as inputs. Multiple steps will run in this step, including `bwa` alignment, mark duplicates and base quality recalibration. GATK ready BAM files will be collected to the output directory.

2. `hapCall` HaplotypeCaller.
The GATK ready BAM and customized json files are inputs in this step. The local paths of GATK bundle files are required to be modified in your json file. A "gVCF" files will be generated.

3. `jdCall` Joint variant discovery
This step will combine the "gVCF" files and then call germline variants in all samples. The paths of the local bundle files are also need to be changed to the json template file. The final VCF file of germline variants will be collected.

An example to run the pipeline.\
<https://hubentu.github.io/others/Rcwl_GATK4.html>

# SessionInfo
```{r}
sessionInfo()
```
